### ISSUE TITLE
`quantile` returns a set for triangular random variables

### ISSUE THREAD

**Issue by peteroupc** [sympy/sympy#21098]
It seems that for triangular random variables (and perhaps for other kinds of random variables), `quantile` returns a set rather than a function.  Perhaps the quantile method should take into account that quantiles of a random variable are increasing functions.

```
In [33]: tri=Triangular('T',0,1,S.Half)

In [34]: qf=quantile(tri)(x)

In [35]: qf                                                                                                                     
Out[35]: 
⎧                                ⎛           ⎧       _______           _______    ⎫⎞            
⎪⎛           ⎧-√2⋅√x   √2⋅√x⎫⎞   ⎜           ⎪  √2⋅╲╱ 1 - x       √2⋅╲╱ 1 - x     ⎪⎟            
⎪⎜[0, 1/2) ∩ ⎨───────, ─────⎬⎟ ∪ ⎜[1/2, 1) ∩ ⎨- ──────────── + 1, ──────────── + 1⎬⎟  for x ≥ 0 
⎨⎝           ⎩   2       2  ⎭⎠   ⎜           ⎪       2                 2          ⎪⎟            
⎪                                ⎝           ⎩                                    ⎭⎠            
⎪                                                                                               
⎩                                        nan                                              otherw

       
       
∧ x ≤ 1
       
       
       
ise    
```

**Comment by gschintgen** [comment]
I've just independently rediscovered this issue with a `ChiSquared` distribution:
```
>>> from sympy.stats import *
>>> X = ChiSquared('X', 2)
>>> quantile(X)(0.95)
{5.99146454710798}
>>> print(_)
FiniteSet(5.99146454710798)
>>>
```
Note the superfluous `FiniteSet`. 

The `Piecewise` below is also very problematic. It's not even consistent: it returns either a set or "numeric" value.
```
>>> quantile(X)(p)
⎧[0, ∞) ∩ {-2⋅log(1 - p)}  for p ≥ 0 ∧ p ≤ 1
⎨
⎩          nan                 otherwise
```

**Comment by gschintgen** [comment]
Had a cursory look at the code:
https://github.com/sympy/sympy/blob/6458e7dd949bcebb397a544db13d6b3b4654f50d/sympy/stats/crv.py#L290-L291

This seems to be a general issue with continuous random variables.

**Comment by oscarbenjamin** [comment]
Maybe it would be better to use `solve` instead of `solveset`.

**Comment by gschintgen** [comment]
I did a few minor experiments and noticed that `solve` isn't able to solve `cdf(tri)(x) - p` where `tri` is the random variable of the OP. `solveset` solves it just fine. The problem with `solveset` is (as always) that it mixes both the actual solutions and their validity conditions in a single large set expression. This makes it difficult to untangle the answers in order to convert them to `Piecewise` expressions.
Also: there's no symbolic infimum for sets. Otherwise we could just write `Infimum(setexpression)` to refer to its single element. (Technically there could be multiple solutions if I'm not mistaken, but then the infimum would still be the one we're looking for.)

**Comment by oscarbenjamin** [comment]
> `solveset` solves it just fine

Yes and no. This is always the conundrum with solveset. In a sense it returns the answer but at the same time it returns something that isn't really usable if you want to do more than just "look at" the output. The contract for solve is that it returns explicit expressions. Having an explicit expression is useful but it isn't possible to make an algorithm that successfully does that in all cases so solve necessarily takes some shortcuts.

In principle solveset can return something more correct but in practice the output is so unstructured that it's hard to make use of even in simple situations. The fact that `quantile` ends up returning a half-set/half-expression mess shows where the mismatch of expectations is between what solveset does and what is actually wanted in this situation.
